code/modules/food_and_drink/cakes.dm

```diff
@@ -24,62 +29,498 @@
+					var/image/buffer = src.GetOverlayImage("[src.overlay_refs[i]]")
+					var/toppingpath = replacetext("[src.overlay_refs[i]]","[replacetext]","slice")
+
+					if((toppingpath == "slice-candle") || (toppingpath == "slice-candle_lit")) //special case for candles :D
+						if(toppingpath == "slice-candle")
+							candle = 1
+						else
+							candle = 2
+							staticiterator--
+							i--
+						src.ClearSpecificOverlays("[src.overlay_refs[i]]")
+						continue
+
+					var/image/toppingimage = new /image('icons/obj/foodNdrink/food_dessert.dmi',toppingpath)
+					if(buffer.color)
+						toppingimage.color = buffer.color
+					cake.UpdateOverlays(toppingimage,toppingpath)
+					src.ClearSpecificOverlays("[src.overlay_refs[i]]")
+					staticiterator--
+					i--
+			else
+				break
+		if(src.amount > 10)
+			src.amount -= 10
+			switch(src.clayer)
+				if(2)
+					src.amount2 = 0
+				if(3)
+					src.amount3 = 0
+			if(mode == CAKE_MODE_CAKE)
+				cake.amount = 10
+			else
+				slices = 10
+		else if((mode == CAKE_MODE_SLICE) && (src.clayer == 1))
+			slices = src.amount
+		else
+			qdel(cake) //this will happen if someone eats a cake without slicing it and the amount math has to recalculate past an entire layer (i.e. someone takes 11 bites)
+			user.show_text("<b>OH NO! The cake was a lie!</b>","red")
+			src.clayer--
+			src.reagents.maximum_volume -= 100
+			return
+
+		src.reagents.maximum_volume -= 100
+		if(mode == CAKE_MODE_SLICE)
+			return list(slices,candle)
+		else
+			src.clayer--
+			src.update_cake_context()
+
+	/*_______________*/
+	/*Light stuffs :D*/
+	/*‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾*/
+
+	proc/ignite(var/mob/user as mob, var/message as text)
+		if (!src)
+			return
+		if (!src.litfam)
+			src.firesource = TRUE
+			src.litfam = TRUE
+			src.hit_type = DAMAGE_BURN
+			src.force = 3
+			src.add_simple_light("cake_light", list(0.5*255, 0.3*255, 0, 100))
+			if (!(src in processing_items))
+				processing_items.Add(src)
+			src.update_cake_context()
+		else
+			return
+
+		if(src.sliced && src.GetOverlayImage("slice-candle"))
+			src.ClearSpecificOverlays("slice-candle")
+			src.UpdateOverlays(image('icons/obj/foodNdrink/food_dessert.dmi',"slice-candle_lit"), "slice-candle_lit")
+		else if(src.GetOverlayImage("cake[src.clayer]-candle"))
+			src.ClearSpecificOverlays("cake[src.clayer]-candle")
+			src.UpdateOverlays(image('icons/obj/foodNdrink/food_dessert.dmi',"cake[src.clayer]-candle_lit"), "cake[src.clayer]-candle_lit")
+		return
+
+
+	proc/put_out(var/mob/user as mob)
+		if (!src) return
+		if (src.litfam)
+			src.firesource = FALSE
+			src.litfam = FALSE
+			hit_type = DAMAGE_BLUNT
+			src.force = 0
+			src.remove_simple_light("cake_light")
+			if (src in processing_items)
+				processing_items.Remove(src)
+			src.update_cake_context()
+		return
+
+	/*__________________*/
+	/*Context Actions :D*/
+	/*‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾*/
+
+	proc/update_cake_context()
+		src.contextActions = list()
```

Comment:
why?

Reply:
it wipes the current context actions and readds them based on the configuration of the cake at that time. This makes sure the buttons are in the same order every time and also removes the buttons that are no longer needed (i.e. if the candle on a cake is blown out, the context actions are refreshed, removing the button to blow out the candle while retaining the button positions.)